index.rst

```diff
@@ -1,51 +1,1166 @@
+        .export_values();
+
+.. warning::
+
+    Do not forget to add the ``.export_values()`` at the end or your enumerated types will not be added to the class!
+
+.. _virtual_functions:
+
+Virtual Functions and abstract Classes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Notice that ``Interpolate`` is an abstract class that cannot be called directly.
+Through examination of ``testInterpolate.py`` we see that ``Interpolate`` objects are created by using the ``makeInterpolate`` function, which is of type ``PTR(Interpolate)``.
+We will wrap ``makeInterpolate`` in :ref:`function_kwargs` but first we finish wrapping ``Interpolate``.
+The main function is the method ``interpolate``, which can be called with a double, list, or ndarray, however calling interpolate with a double is actually a virtual function, so we cannot wrap it directly.
+
+Instead we create a lambda function:
+
+.. code-block:: c++
+
+    clsInterpolate.def("interpolate", [](Interpolate &t, double const x) -> double {
+            return t.interpolate(x);
+    });
+
+This defines the function ``Interpolate::interpolate``, which then calls the virtual function ``interpolate`` of the ``Interpolate`` object directly (the method exists, just not in a way that it can be wrapped by pybind11).
+
+NDArray's
+^^^^^^^^^
+
+Since the ``interpolate`` method is an overloaded function, only one of which is virtual, we can wrap the other function definitions in the traditional way:
+
+.. code-block:: c++
+
+    clsInterpolate.def("interpolate",
+                       (std::vector<double> (Interpolate::*) (std::vector<double> const&) const)
+                           &Interpolate::interpolate);
+    clsInterpolate.def("interpolate",
+                       (ndarray::Array<double, 1> (Interpolate::*) (ndarray::Array<double const, 1> const&)
+                           const) &Interpolate::interpolate);
+
+However, since we are using ndarray's we also need to include the numpy and ndarray headers at the top of ``interpolate.cc``
+
+.. code-block:: c++
+
+    #include "numpy/arrayobject.h"
+    #include "ndarray/pybind11.h"
+    #include "ndarray/converter.h"
+
+It is also necessary to check that numpy has been installed and setup (otherwise unexpected segfaults will occcur), so in the module definition we add
+
+.. code-block:: c++
+
+    if (_import_array() < 0) {
+            PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import");
+            return nullptr;
+        }
+
+.. _function_kwargs:
+
+Wrapping Functions with Default Arguments
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The final method remaining to wrap in ``interpolate.h`` is ``makeInterpolate``, which creates an ``Interpolate`` object from the virtual class.
+
+This is an overloaded function, so we define it in the usual way but add ``py::arg("paremeter")`` for *all* of the arguments of the function (not just the ones that we need to give default values).
```

Comment:
Perhaps. Pim and I weren't adding them but I noticed that Russel make extensive use of them in his code and I agree that can be helpful.